### PR TITLE
Design and implement ir to represent program

### PR DESCRIPTION
**Pull request by mayanksharma3**
### Clubhouse Ticket
[ch47](https://app.clubhouse.io/manim-dsl/story/47/design-and-implement-ir-to-represent-program-state-and-code-line)

### Description

Added the internal representation and code generation step to produce manim code and a python file. Also added a ManimWriter to save those files and also create the mp4 by calling manim

### Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### How Has This Been Tested?

Please describe tests added as well.

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by mayanksharma3** [comment]
Looks all ready to merge

**Comment by arj119** [comment]
> Looks all ready to merge

I will probably need to modify the test since @iz0825 changes to the python @mayanksharma3

File: src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+/** Objects **/
+
+interface Object : ManimInstr {
+    val ident: String
+}
+
+enum class ObjectSide(var coord: Pair<Double, Double>) {
+    ABOVE(Pair(0.0, 0.25)),
+    BELOW(Pair(0.0, 0.25)),
```
Comment (claralebbos) [inline src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt:13]: 0.0 -0.25?

File: src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    LEFT(Pair(0.0, 0.25)),
```
Comment (claralebbos) [inline src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt:14]: -0.25 0.0

File: src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    RIGHT(Pair(0.0, 0.25));
```
Comment (claralebbos) [inline src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt:15]: 0.25 0?

**Comment by mayanksharma3** [comment]
@arj119 Could you remove "org.apache.commons:commons-io:1.3.2" and use java files instead?

Like `File(...).readText()` gives the file as a string

**Comment by mayanksharma3** [comment]
> @arj119 Could you remove "org.apache.commons:commons-io:1.3.2" and use java files instead?
> 
> Like `File(...).readText()` gives the file as a string

I can do this in a bit if there's a break in the lecture if needed

**Comment by mayanksharma3** [comment]
I've updated the ObjectSide and the tests to not use apache now, but as @claralebbos put on slack we need to add the instruction code as well to fade out/remove a shape

**Comment by adityagoel4512** [comment]
> I've updated the ObjectSide and the tests to not use apache now, but as @claralebbos put on slack we need to add the instruction code as well to fade out/remove a shape

pop() result should probably only fade out if the return value is not assigned to a variable. Can the IR be adapted to make that happen first. @mayanksharma3

**Comment by mayanksharma3** [comment]
The instruction itself could have a flag for whether it should fade out or not? And if not then the next instruction (which will show like "x = ...." will take the rectangle that has been popped off and move it to the text

**Comment by adityagoel4512** [comment]
> The instruction itself could have a flag for whether it should fade out or not? And if not then the next instruction (which will show like "x = ...." will take the rectangle that has been popped off and move it to the text

Yes I suppose we could show the state of all the current variables looking forward and visualise the number going to the relevant variable. We will have to wait for VM for proper implementation but flag makes sense.

**Comment by mayanksharma3** [comment]
> > The instruction itself could have a flag for whether it should fade out or not? And if not then the next instruction (which will show like "x = ...." will take the rectangle that has been popped off and move it to the text
> 
> Yes I suppose we could show the state of all the current variables looking forward and visualise the number going to the relevant variable. We will have to wait for VM for proper implementation but flag makes sense.

Yes the VM is now merged in, and each step we can tell the state of each variable, so we could even show that above the code at each step

**Comment by adityagoel4512** [comment]
> > > The instruction itself could have a flag for whether it should fade out or not? And if not then the next instruction (which will show like "x = ...." will take the rectangle that has been popped off and move it to the text
> > 
> > 
> > Yes I suppose we could show the state of all the current variables looking forward and visualise the number going to the relevant variable. We will have to wait for VM for proper implementation but flag makes sense.
> 
> Yes the VM is now merged in, and each step we can tell the state of each variable, so we could even show that above the code at each step

I was going to add that ASTExecutor.kt is still being spiked and so shouldn't end up on master just now. There will be a new branch for work on it and I think we should plan it out more thoroughly first.

**Comment by adityagoel4512** [comment]
The plan for spikes was to upload onto Clubhouse first right

**Comment by adityagoel4512** [comment]
I think just a flag in the IR will do for now. Is that okay with you @mayanksharma3 ?

**Comment by mayanksharma3** [comment]
> The plan for spikes was to upload onto Clubhouse first right

It was spiked and then implemented with tests as we need it to get the IR generated in the first place so we need some implementation which works for our current language (which it does)

**Comment by iz0825** [comment]
> > I think just a flag in the IR will do for now. Is that okay with you @mayanksharma3 ?
> 
> Yes it will, I am not free till 4PM though because of lectures and UTA so @arj119 @iz0825 if one of you could do it that would be great

I will try and maybe you can check if what I did is what we wanted later when you are free

**Comment by mayanksharma3** [comment]
Yup looks good to me, thanks @iz0825 !

**Review by rbanerjee20 — APPROVED** [review]

File: src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+/** Objects **/
+
+interface Object : ManimInstr {
+    val ident: String
+}
+
+enum class ObjectSide(var coord: Pair<Double, Double>) {
+    ABOVE(Pair(0.0, 0.25)),
+    BELOW(Pair(-0.25, 0.0)),
```
Comment (claralebbos) [inline src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt:13]: shouldn't this be 0.0, -0.25?

File: src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    LEFT(Pair(-0.25, 0.0)),
+    RIGHT(Pair(0.025, 0.0));
```
Comment (claralebbos) [inline src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt:15]: 0.25, 0.0

File: src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt (hunk 4 above)
Comment (arj119) [reply src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt:13]: @iz0825 ?

File: src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt (hunk 5 above)
Comment (arj119) [reply src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt:15]: @iz0825 ?

Comment (arj119) [reply src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt:15]: Will address later

File: src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt (hunk 4 above)
Comment (arj119) [reply src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt:13]: Will also address later

File: src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt (hunk 5 above)
Comment (iz0825) [reply src/main/kotlin/com/manimdsl/linearrepresentation/Objects.kt:15]: I think clara is right i'll fix it and push again

**Pull request merged**